SignalShareExtension/ShareViewController.swift

```diff
@@ -459,13 +485,82 @@ public class ShareViewController: UINavigationController, ShareViewDelegate, SAE
             if url.pathExtension.count > 0 {
                 // Determine a more specific utiType based on file extension
                 if let typeExtension = MIMETypeUtil.utiType(forFileExtension: url.pathExtension) {
+                    Logger.debug("\(self.logTag) utiType based on extension: \(typeExtension)")
                     specificUTIType = typeExtension
                 }
             }
 
-            let attachment = SignalAttachment.attachment(dataSource: dataSource, dataUTI: specificUTIType, imageQuality:.medium)
+            guard !SignalAttachment.isInvalidVideo(dataSource: dataSource, dataUTI: specificUTIType) else {
+                // This can happen, e.g. when sharing a quicktime-video from iCloud drive.
+
+                let (promise, exportSession) = SignalAttachment.compressVideoAsMp4(dataSource: dataSource, dataUTI: specificUTIType)
+
+                // TODO: How can we move waiting for this export to the end of the share flow rather than having to do it up front?
+                // Ideally we'd be able to start it here, and not block the UI on conversion unless there's still work to be done
+                // when the user hit's "send".
```

Comment:
"You can solve every problem with another level of indirection, except for the problem of too many levels of indirection."

* `SignalAttachment` is itself a layer of indirection around a `DataSource`. 
* `DataSource` is a layer of indirection (adapter pattern) around a file or NSData.

I guess we could have a `SignalAttachmentSource`, but IMO the complexity isn't worth it.